source/portable/NetworkInterface/TMS320F2838x/NetworkInterface.c

```diff
@@ -0,0 +1,304 @@
+
+    //
+    //Change the priority of the interrupt handlers
+    //
+
+    Interrupt_setPriority(INT_EMAC_TX0, 15);
+    Interrupt_setPriority(INT_EMAC_RX0, 15);
+//    Interrupt_setPriority(INT_EMAC, 15);
+
+    //
+    //Enable the default interrupt handlers
+    //
+    Interrupt_enable(INT_EMAC_TX0);
+    Interrupt_enable(INT_EMAC_RX0);
+//    Interrupt_enable(INT_EMAC);
+
+    if(pInitCfg == NULL){
+        xReturn = pdFAIL;
+    }else{
+        NetworkEndPoint_t* pxEndPoint = FreeRTOS_FirstEndPoint(pxInterface);
+        uint8_t MACAddress[ipMAC_ADDRESS_LENGTH_BYTES] = pxEndPoint->xMACAddress.ucBytes;
+        uint32_t MACAddressLow = (pxEndPoint->xMACAddress.ucBytes[0]) | (pxEndPoint->xMACAddress.ucBytes[1] << 8) | (pxEndPoint->xMACAddress.ucBytes[2] << 16) | (pxEndPoint->xMACAddress.ucBytes[3]<<24);
+        uint32_t MACAddressHigh = (pxEndPoint->xMACAddress.ucBytes[4]) | (pxEndPoint->xMACAddress.ucBytes[5] << 8);
+        Ethernet_setMACAddr(EMAC_BASE, 0, MACAddressHigh, MACAddressLow, ETHERNET_CHANNEL_0);
+        xReturn = pdPASS;
+    }
+
+    if(xEMACTaskHandle == NULL){
+        xTaskCreate(prvEMACDeferredInterruptHandlerTask, "EMACInt", configMINIMAL_STACK_SIZE, NULL, configMAX_PRIORITIES - 1, &xEMACTaskHandle);
+        configASSERT(xEMACTaskHandle);
+        pxMyInterface = pxInterface;
+    }
+
+    return xReturn;
+}
+
+static BaseType_t prvf28388d_NetworkInterfaceOutput(NetworkInterface_t* pxInterface, NetworkBufferDescriptor_t* const pxDescriptor, BaseType_t xReleaseAfterSend){
+    pxGlobalDescriptor = pxDescriptor;
+    
+    Ethernet_Pkt_Desc pktDesc;
+    
+    pktDesc.bufferLength = pxDescriptor->xDataLength;
+    pktDesc.dataOffset = 0;
+    pktDesc.dataBuffer = pxDescriptor->pucEthernetBuffer;
+    pktDesc.nextPacketDesc = 0;
+    pktDesc.flags = ETHERNET_PKT_FLAG_SOP | ETHERNET_PKT_FLAG_EOP | ETHERNET_PKT_FLAG_SA_INS | ETHERNET_PKT_FLAG_SAIC;
+    pktDesc.pktChannel = ETHERNET_DMA_CHANNEL_NUM_0;
+    pktDesc.pktLength = pxDescriptor->xDataLength;
+    pktDesc.validLength = pxDescriptor->xDataLength;
+    pktDesc.numPktFrags = 1;
+
+    uint32_t ret = Ethernet_sendPacket(emac_handle, &pktDesc);
+
+    if(ret == 0){
+        vReleaseNetworkBufferAndDescriptor(pxDescriptor);
+    }
+
+    if(xReleaseAfterSend != pdFALSE){
+        pxDescriptor->pucEthernetBuffer = NULL;
+        vReleaseNetworkBufferAndDescriptor(pxDescriptor);
+    }
+}
+
+static void prvEMACDeferredInterruptHandlerTask( void *pvParameters )
+{
+    NetworkBufferDescriptor_t *pxBufferDescriptor;
+    IPStackEvent_t xRxEvent;
+    uint32_t ulISREvents = 0U;
+    for( ;; )
+    {
+        xTaskNotifyWait( 0U,
+                         EMAC_IF_ALL_EVENT,
+                         &( ulISREvents ),
+                         portMAX_DELAY );
+        if((ulISREvents & EMAC_IF_RX_EVENT) != 0){
+            pxBufferDescriptor = pxGetNetworkBufferWithDescriptor( ETHERNET_MAX_PACKET_LENGTH, 0 );

```

Comment:
Here you are handling a `EMAC_IF_RX_EVENT`, meaning that a packet has been received. Wouldn't you check the validity of `pPktDescGlobal` here, see if `validLength` is within range and so?
Not need to copy `ETHERNET_MAX_PACKET_LENGTH` bytes, that would be a waste of time.
Also I wouldn't create a packet buffer before calling `eConsiderFrameForProcessing()` and `FreeRTOS_MatchingEndpoint()`: both functions might fail.
Have a look at this:
~~~
if( valid_packet( pPktDescGlobal )
{
    /* Check if we should handle this packet. */
    if( eConsiderFrameForProcessing( pPktDescGlobal->dataBuffer ) == eProcessBuffer )
    {
        /* Find the proper endpoint. */
        NetworkEndPoint_t * pxEndPoint = FreeRTOS_MatchingEndpoint( pxMyInterface, pPktDescGlobal->dataBuffer );

        if( pxEndPoint != NULL )
        {
            NetworkBufferDescriptor_t * pxBufferDescriptor = pxGetNetworkBufferWithDescriptor( pPktDescGlobal->validLength, 0 );
            memcpy( pxBufferDescriptor->pucEthernetBuffer, pPktDescGlobal->dataBuffer, pPktDescGlobal->validLength );
            pxBufferDescriptor->xDataLength = pPktDescGlobal->validLength;

            pxBufferDescriptor->pxEndPoint = pxEndPoint;
            xRxEvent.eEventType = eNetworkRxEvent;
            xRxEvent.pvData = ( void * ) pxBufferDescriptor;

            if( xSendEventStructToIPTask( &xRxEvent, 0 ) == pdTRUE )
            {
                iptraceNETWORK_INTERFACE_RECEIVE();
            }
            else
            {
                iptraceETHERNET_RX_EVENT_LOST();
                vReleaseNetworkBufferAndDescriptor( pxBufferDescriptor );
            }
        }
        else
        {
            iptraceETHERNET_RX_EVENT_LOST();
        }
    }
    else
    {
        iptraceETHERNET_RX_EVENT_LOST();
    }
}
~~~
In other words, do not call `pxGetNetworkBufferWithDescriptor()` until you are sure that it will be used.

Question: How do we make sure that `pPktDescGlobal` isn't accessed before by interrupt code before we've handled the last packet?
I am thinking of adding a **queue** or a **counting semaphore** between the EMAC task and the callback functions:
~~~
Ethernet_receivePacketCallbackCustom()
Ethernet_releaseTxPacketBufferCallbackCustom()
~~~
A `EMAC_IF_RX_EVENT` event means: one or more packets have been received.
A `EMAC_IF_TX_EVENT` event means the one or more packets have been sent.
But while you are handling some `EMAC_IF` event, the interrupt may get triggered.